astropy/io/votable/tests/test_vo.py

```diff
@@ -276,7 +279,87 @@ def test_select_columns_by_name():
     assert array["string_test"][0] == "String & test"
     for c in columns:
         assert not np.all(mask[c])
-    assert np.all(mask["unicode_test"])
+
+    # deselected columns shouldn't be present in the output
+    assert "unicode_test" not in array.dtype.fields
+    assert "unicode_test" not in mask.dtype.fields
+
+
+@pytest.mark.parametrize(
+    "column_ids, use_names_over_ids, expected_names",
+    [
+        # just the first column
+        pytest.param(
+            ["string_test"],
+            False,
+            ["string_test"],
+            id="first-col-ids",
+        ),
+        pytest.param(
+            ["string_test"],
+            True,
+            ["string test"],
+            id="first-col-names",
+        ),
+        # a single column, other than the first
+        pytest.param(
+            ["unicode_test"],
+            False,
+            ["unicode_test"],
+            id="single-col-ids",
+        ),
+        pytest.param(
+            ["unicode_test"],
+            True,
+            ["unicode_test"],
+            id="single-col-names",
+        ),
+        # two non-consecutive, differently named columns
+        pytest.param(
+            ["string_test", "unicode_test"],
+            False,
+            ["string_test", "unicode_test"],
+            id="two-cols-ids",
+        ),
+        pytest.param(
+            ["string_test", "unicode_test"],
+            True,
+            ["string test", "unicode_test"],
+            id="two-cols-names",
+        ),
+        # just the first two columns (that have the same ID)
+        pytest.param(
+            ["string_test", "string_test_2"],
+            False,
+            ["string_test", "string_test_2"],
+            id="two-cols-ids-sameID",
+        ),
+        pytest.param(
+            ["string_test", "string_test_2"],
+            True,
+            ["string test", "fixed string test"],
+            id="two-cols-names-sameID",
+        ),
+    ],
+)
+def test_select_columns_by_name_edge_cases(
+    column_ids, use_names_over_ids, expected_names
+):
+    # see https://github.com/astropy/astropy/issues/14943
+    filename = get_pkg_data_filename("data/regression.xml")
+    with np.errstate(over="ignore"):
+        # https://github.com/astropy/astropy/issues/13341
+        vot1 = parse_single_table(filename, columns=column_ids)
+    t1 = vot1.to_table(use_names_over_ids=use_names_over_ids)
+    assert t1.colnames == expected_names
+
+    # columns should be returned in the order they are found, which
+    # in the general case isn't the order they are requested
+    with np.errstate(over="ignore"):
+        # https://github.com/astropy/astropy/issues/13341
+        vot2 = parse_single_table(filename, columns=list(reversed(column_ids)))
```

Comment:
For half the tests `column_ids` only has a single element, so there's little reason to reverse it. It looks like having separate tests for single-element and two-element `column_ids` could simplify parametrization a lot.

Reply:
I think there's value in keeping the current parametrisation, which makes it trivial to select test cases related to this (otherwise untested) feature. Honestly I don't really see what problem would be solved by splitting the test in two. Is it okay if I just return early when this part of the test is redundant ?
```patch
diff --git a/astropy/io/votable/tests/test_vo.py b/astropy/io/votable/tests/test_vo.py
index c65189e1f2..712191828d 100644
--- a/astropy/io/votable/tests/test_vo.py
+++ b/astropy/io/votable/tests/test_vo.py
@@ -353,6 +353,9 @@ def test_select_columns_by_name_edge_cases(
     t1 = vot1.to_table(use_names_over_ids=use_names_over_ids)
     assert t1.colnames == expected_names
 
+    if len(column_ids) < 2:
+        return
+
     # columns should be returned in the order they are found, which
     # in the general case isn't the order they are requested
     with np.errstate(over="ignore"):
```

Reply:
If you name the two tests I'm proposing well enough then they can still be easily selected with [`pytest -k ...`](https://docs.pytest.org/en/stable/how-to/usage.html#specifying-which-tests-to-run)

Reply:
sure, but then I'll essentially just duplicate the test code entirely. Is it worth it ?

Reply:
For the single-element test the parametrization would be so much simpler that it should be worth it.

Reply:
I'm honestly not sure, but I'll push the change you're proposing as a separate commit so we know for sure if we're talking about the same thing.

Reply:
I thought that in the single-element case `column_id` is always the same as `expected_name`, but I just noticed that in one case the `column_id` is `"string_test"`, but the `expected_name` is `"string test"`. So the simplifications I had in mind are not possible.

It is possible to have a single test without reversing `column_ids` in its code by providing the reversed `column_ids` as additional parameters.

Reply:
>  So the simplifications I had in mind are not possible.

I went through the same process while trying to implement your suggestion actually :)

> It is possible to have a single test without reversing column_ids in its code by providing the reversed column_ids as additional parameters.

this seems reasonable to me. Done !